# Visualizations in Python
### Using DJ Data

#### Imports

In [51]:
import pandas as pd
import plotly.express as px

### Part 1 - Loungers/Quiet Type - columns high 'acousticness', low 'energy', low 'loudness'

#### Lowest Energy Songs

In [52]:
# reading the csv file
# looking at columns energy, loudness, and acousticness
df = pd.read_csv('music_genre_classification (1).csv', usecols = ['Artist Name', 'energy', 'loudness', 'acousticness'])

In [53]:
# only getting the top 10 smallest energy values
low_energy = df.nsmallest(10, 'energy')
print(low_energy[['Artist Name', 'energy']])

                                            Artist Name   energy
6089                               Todd Anderson-Kunert  0.00002
2801         Ludwig van Beethoven, Emanuel Ax, Yo-Yo Ma  0.00433
1521  G. I. Gurdjieff, Anja Lechner, Vassilis Tsabro...  0.00526
7432                                              HILMA  0.00700
3561  Luigi Boccherini, Tim Hugh, Scottish Chamber O...  0.00714
1985                              Gerry Mulligan Sextet  0.00894
2919                                 Broccoli, you too?  0.00976
3373             Francisco Villegas, Julio Mariñelarena  0.01040
5833                                      Library Tapes  0.01080
3997                                        Carlos Cipa  0.01230


In [54]:
# energy barchart - data is from filtered top 10 smallest energy values above
px.bar(low_energy, 'Artist Name', 'energy', title = '10 Lowest Energy Songs by Artist', color = 'energy', color_continuous_scale = 'Greens')

#### Lowest Loudness Rating (decibels)

In [55]:
# lowest 10 loudness values in decibels
quietest = df.nsmallest(10, 'loudness')
print(quietest[['Artist Name', 'loudness']])

                                            Artist Name  loudness
3561  Luigi Boccherini, Tim Hugh, Scottish Chamber O...   -36.544
2801         Ludwig van Beethoven, Emanuel Ax, Yo-Yo Ma   -32.639
1822  Alexander Glazunov, Alexander Rudin, Moscow Ph...   -32.520
3150                      Victor Kozlov, Irina Kulikova   -31.846
1268  Johann Sebastian Bach, Freiburger Barockorchester   -30.625
7432                                              HILMA   -29.955
1521  G. I. Gurdjieff, Anja Lechner, Vassilis Tsabro...   -29.845
6651              Ottorino Respighi, Magdalena Hoffmann   -29.234
3997                                        Carlos Cipa   -29.161
4895                                      Tony Anderson   -28.958


In [56]:
# loudness bar chart
px.bar(quietest, 'Artist Name', 'loudness', title = '10 Quietest Songs by Artist', color = 'loudness', color_continuous_scale = 'blues')

#### Highest 10 Acoustic Values

In [57]:
# highest 10 acoustic values
best_acoustics = df.nlargest(10, 'acousticness')
print(best_acoustics[['Artist Name', 'acousticness']])

                           Artist Name  acousticness
757        Memphis Minnie & Kansas Joe         0.995
1264                  Blind Boy Fuller         0.995
3206                   Baby Boy Warren         0.995
5244  Anja Lechner, François Couturier         0.995
6349                     Raquel Meller         0.995
1404                  Daniel Raticelli         0.994
2124                     Juanita Reina         0.994
3649                   Thelonious Monk         0.994
4237              Antonio Garcia Isaac         0.994
6154                        R.J. Smith         0.994


In [58]:
px.bar(best_acoustics, 'Artist Name', 'acousticness', title = '10 Best Acoustic Songs by Artist', color = 'acousticness', color_continuous_scale = 'sunsetdark')

Question #12) What do you notice about this bar chart that's different from the others? Why do you think it's different?

With the nlargest, the bar graph shows approximately the same bar height for each artist since they are all relatively close in acousticness, and instead shows the acoustic level in the form of different colors. The reason for this is because we chose to use the color scale 'sunsetdark'.

#### Changing the 10 value for nlargest to 30

In [59]:
# using 30 instead of 10
best_acoustics_ = df.nlargest(30, 'acousticness') # will give us the top 30 largest instead of top 10
print(best_acoustics_[['Artist Name', 'acousticness']])

                                      Artist Name  acousticness
757                   Memphis Minnie & Kansas Joe         0.995
1264                             Blind Boy Fuller         0.995
3206                              Baby Boy Warren         0.995
5244             Anja Lechner, François Couturier         0.995
6349                                Raquel Meller         0.995
1404                             Daniel Raticelli         0.994
2124                                Juanita Reina         0.994
3649                              Thelonious Monk         0.994
4237                         Antonio Garcia Isaac         0.994
6154                                   R.J. Smith         0.994
6426  Franz Schubert, Mischa Maisky, Daria Hovora         0.994
6700                                 Diane Edison         0.994
630                            Juanito Valderrama         0.993
688                               Lata Mangeshkar         0.993
893             Rekha, Vinod Khanna, Ram

In [60]:
px.bar(best_acoustics_, 'Artist Name', 'acousticness', title = '30 Best Acoustic Songs by Artist', color = 'acousticness', color_continuous_scale = 'sunsetdark')

Question #13) Is it more clear why the bar chart was different? Explain.

Viewing the top 30 instead of top 10 provides a wider range of colors to discern the acousticness of each artist.

### Part 2: Create Artist List and Visualize for the Other Two Audience Types

In [61]:
# repeat for track team/partiers & NYT bestseller book club/literary fans
# code will be nearly identical but take note of...
    # column names (so read csv file again but with correct column names for traits of each type
    # nsmallest vs nlargest
    # choose your own colors (color_continuous_scale = 'your_choice_of_color')
    # meaningful trait variable names
    # each time you reread csv file you make a copy of the variable df (that's why you can use it again for a new selection of columns)

### Part 2a: Track Team/Partiers - columns high 'energy', high 'tempo', & short 'duration_in min/ms'

In [62]:
# reading csv file
# looking at 'Artist Name', 'energy', 'tempo', 'duration_in min/ms'
df = pd.read_csv('music_genre_classification (1).csv', usecols = ['Artist Name', 'energy', 'tempo', 'duration_in min/ms'])

In [63]:
high_energy = df.nlargest(30, 'energy')
print(high_energy[['Artist Name', 'energy']])

                  Artist Name  energy
370                Darkthrone   1.000
15                  FEVER 333   0.998
4241                FEVER 333   0.998
6208                     Lard   0.998
1387           Pretty Vicious   0.997
1650               Hellripper   0.997
4934               Hellhammer   0.997
5197        The Acacia Strain   0.997
283             Ozzy Osbourne   0.996
959              Funeral Mist   0.996
1047       Laurenne / Louhimo   0.996
1258              Destruction   0.996
2870             The Northern   0.996
3332           A Lethal Smile   0.996
3396               Stone Sour   0.996
3769                   Hatena   0.996
4197                Vengeance   0.996
4966  Five Finger Death Punch   0.996
6966              The Go-Go's   0.996
7436         Four Year Strong   0.996
1633  Bullet For My Valentine   0.995
2085                  Polaris   0.995
5381            Still Remains   0.995
5468         August Burns Red   0.995
6446                Gama Bomb   0.995
6521        

In [64]:
px.bar(high_energy, 'Artist Name', 'energy', title = 'Top 30 Highest Energy Songs by Artist', color = 'energy', color_continuous_scale = 'icefire')

#### Top 30 Highest Tempo Songs by Artist

In [65]:
# view top 30 highest tempo songs
high_tempo = df.nlargest(30, 'tempo')
print(high_tempo[['Artist Name', 'tempo']])

                                            Artist Name    tempo
3581                                                IDK  249.438
2579                                         New Candys  216.194
5150                                         New Candys  216.194
4424                                         frown line  216.091
7018                                  Angels & Airwaves  216.020
5465                                    Ólafur Arnalds  212.849
4385                                           AFN Peso  212.166
2390                                               Evil  209.657
94                                          Aisha Badru  208.648
3688                                      Johnny Winter  208.576
7379                                  Souls Of Mischief  206.247
1279                                     Dierks Bentley  205.972
5634                            Scarecrow Blues Hip Hop  205.956
6931                                            Rihanna  205.846
3379                     

In [66]:
px.bar(high_tempo, 'Artist Name', 'tempo', title = 'Top 30 Highest Tempo Songs by Artist', color = 'tempo', color_continuous_scale = 'emrld')

#### Top 30 Shortest Songs by Artist

In [67]:
# view top 30 shortest songs
short = df.nsmallest(30, 'duration_in min/ms')
print(short[['Artist Name', 'duration_in min/ms']])

                                            Artist Name  duration_in min/ms
4010                                    Mahendra Kapoor            0.687450
2132                                       R. D. Burman            1.209367
4066                         Ilaiyaraaja, K. J. Yesudas            1.291667
5974               Eleni Karaindrou, Alexandros Botinis            1.353333
6447                               Kavita Krishnamurthy            1.426883
725                                       Reyanna Maria            1.429033
4274                                       R. D. Burman            1.434450
609                                        Bappi Lahiri            1.464217
5438              Mani Sharma, Amit Heri, Rajesh Vaidya            1.495367
7561                                           Labrinth            1.531167
2109                                     Applewood Road            1.608017
2475                                     PinkPantheress            1.642400
3635        

In [68]:
# duration bar graph
px.bar(short, 'Artist Name', 'duration_in min/ms', title = 'Top 30 Shortest Songs by Artist', color = 'duration_in min/ms', color_continuous_scale = 'solar')

### Part 2b) NYT bestseller book club/literary fans - columns high 'Popularity', high 'danceability', & high 'speechiness'

In [69]:
# read csv for columns 'Popularity', 'danceability', & 'speechiness'
df = pd.read_csv('music_genre_classification (1).csv', usecols = ['Artist Name', 'Popularity', 'danceability', 'speechiness'])

#### Top 20 Most Popular Songs by Artist

In [70]:
high_pop = df.nlargest(20, 'Popularity')
print(high_pop, [['Artist Name', 'Popularity']])

                  Artist Name  Popularity  danceability  speechiness
6062                Lil Nas X       100.0         0.610       0.1520
3797            Justin Bieber        99.0         0.677       0.1190
2359                 Doja Cat        98.0         0.762       0.0286
1816              Masked Wolf        97.0         0.778       0.0913
6736               Ed Sheeran        97.0         0.808       0.0348
3315               Bruno Mars        96.0         0.586       0.0324
4263                Bad Bunny        96.0         0.644       0.1180
6038  J Balvin, Maria Becerra        95.0         0.891       0.1060
777                The Weeknd        94.0         0.514       0.0598
4190            Billie Eilish        93.0         0.634       0.0807
6226           Olivia Rodrigo        92.0         0.556       0.2040
7206   Lil Nas X, Jack Harlow        92.0         0.736       0.0615
740                  Maroon 5        91.0         0.713       0.0270
2966                 Dua Lipa     

In [71]:
px.bar(high_pop, 'Artist Name', 'Popularity', title = 'Top 20 Most Popular Songs by Artist', color = 'Popularity', color_continuous_scale = 'twilight')

#### Top 20 Highest Danceable Songs by Artist

In [72]:
high_danceablility = df.nlargest(20, 'danceability')
print(high_danceablility[['Artist Name', 'danceability']])

                               Artist Name  danceability
2326                            Rebelution         0.975
4120                            Gucci Mane         0.972
7116                                  Tyga         0.968
3868                           Jody Breeze         0.966
4930                          Run–D.M.C.         0.965
4019              Ashnikko, Princess Nokia         0.964
4215                              Ashnikko         0.964
1576                             Boy 8-Bit         0.962
5417                            Lou Hayter         0.962
6070                                John K         0.958
40                               Tim Atlas         0.957
7124                   Megan Thee Stallion         0.957
7600                   Megan Thee Stallion         0.957
6142                           Erica Banks         0.956
1118                                 Trina         0.955
2627                            Asian Doll         0.955
1612                    Kardina

In [73]:
px.bar(high_danceablility, 'Artist Name', 'danceability', title = 'Top 20 Most Danceable Songs by Artist', color = 'danceability', color_continuous_scale = 'electric')

#### Top 20 Highest Speechiness Songs by Artist

In [74]:
speech = df.nlargest(20, 'speechiness')
print(speech[['Artist Name', 'speechiness']])

                                            Artist Name  speechiness
1043                                       Ishmael Reed        0.960
5763                                       Amiri Baraka        0.938
7005  Rajesh Khanna, Sumita Sanyal, Ramesh Deo, Seem...        0.829
734                                    Amitabh Bachchan        0.814
4740                                    Langston Hughes        0.806
580                                     Margaret Walker        0.678
1602                                           Rob Frye        0.667
3646                                          Son House        0.658
4446                                       Moneybagg Yo        0.619
893                   Rekha, Vinod Khanna, Ram P. Sethi        0.565
6975                                            DijahSB        0.562
1962                                             Cordae        0.558
4864                                           Mr. Oizo        0.550
3744                              

In [75]:
px.bar(speech, 'Artist Name', 'speechiness', title = 'Top 20 Highest Speechiness Songs by Artist', color = 'speechiness', color_continuous_scale = 'rainbow')

#### More Thought:

Q #1)


A. Look closely at the track team charts for energy and tempo. There's a duplicate in each. How does that interrupt your bar chart?

For 'energy', there are duplicates for the artists 'FEVER 333' and 'Store Sour'. The result of the bar chart shows the two songs by the same band as an additional bar above each. 'FEVER 333' has two songs in the top 30 and show approximately the same values for energy, but 'Sour Stone' has two songs with two different energy values. For 'tempo', the artist 'New Candys' are repeated since they have two songs that qualify for the top 30 highest tempo songs. The result is a bar graph with two bars, one stacked on the other.


B. What should have been done with the data before plotting it to avoid this?

For these bar graphs the stacked bars need to be separated and each song needs to be separated by values. The solution to this problem would be to separate each song so that there are two separate bars and have each artist name be repeated on the x-axis by song.


Q #2) After you complete the bar charts for all three audience types, look for overlap. For example, what artists are present in all three charts for the lounger type? Write those down as must-have songs for your song list.

Loungers/Quiet Type repeats:
Carlos Cipa, G. I. Gurdjieff (& others), HILMA, Ludwig van Beethoven (& others), Luigi Boccerini are reapted on the quietest and lowest energy bar charts

Track Team/Partiers repeats:
There are no repeat artists between bar graphs for track team/partiers.

NYT bestseller book club/Literary fans repeats:
There are no repeat artists between bar graphs for NYT bestseller book club/literary fans.


